### PR TITLE
Changes to allow a per-cluster image

### PR DESCRIPTION
**Pull request by mkjpryor-stfc**

### CODE REVIEW

**Review by oneswig — APPROVED** [review]
Looks workable to me. It's too bad there don't appear to be any Heat resource templates for Glance images, which could have neatly encapsulated the image and it's management within the stack.

It might be a good idea to use something more unique than cluster name (eg, uuid), given that users in a large project might create similarly-named cluster deployments, and confusion could ensue.

**Comment by mkjpryor-stfc** [comment]
@oneswig 

There is one Heat primitive, but it only lets you create an image from an ISO by URL.

The image creation has to happen before the stack creation, so I’m afraid cluster name is the best we can do. I don’t particularly like a solution that involves modifying the inventory after the first run to inject an image UUID. I’m not even sure how I would do it from Ansible...

You’ll notice that the images get a property that identifies them as a cluster image - I’ve written a patch for the portal that excludes these images from the list displayed to the user so portal users won’t see these images as an option for creating ad-hoc machines.  I hoped that would avoid most of the confusion.

**Comment by mkjpryor-stfc** [comment]
@oneswig Any more comments, or are you happy to merge this?

**Comment by brtkwr** [comment]
What about a cleanup step that deletes the volume when the cluster is deleted?

**Comment by mkjpryor-stfc** [comment]
@brtknr

You don't need to wait for cluster deletion: this step deletes the volume once the image has been uploaded - https://github.com/cedadev/jasmin-appliances/pull/63/files#diff-ad24ad2fb5d8f31b6b29a53301a806fcR90

And this step removes the image when the cluster is deleted: https://github.com/cedadev/jasmin-appliances/pull/63/files#diff-ad24ad2fb5d8f31b6b29a53301a806fcR105

File: tasks/infra/provision.yml [hunk 1]
```diff
@@ -45,9 +46,64 @@
...
   include_role:
     name: jasmin.cluster-infra
   vars:
     cluster_groups: "{{ cluster_infra_groups | default([]) }}"
+
+- name: Remove cluster image
+  os_image:
+    name: "cluster-{{ cluster_name }}"
+    state: absent
+  when: cluster_state | default('present') | lower != 'present'
```
Comment (brtkwr) [inline tasks/infra/provision.yml:109]: isn't it better explicitly to say cluster_state == 'absent' here?

Comment (mkjpryor-stfc) [reply tasks/infra/provision.yml:109]: I don't mind either way. I sort of like the idea of thinking of being present or not present, but have no strong opinion. There is nowhere else that checks this that I can find.

present/absent being the only choices should be enforced by other modules.

**Pull request merged**